src/v/cluster/members_manager.cc

```diff
@@ -456,59 +459,68 @@ auto members_manager::dispatch_rpc_to_leader(
+      req.node.id(),
+      req.logical_version);
+
     // curent node is a leader
     if (_raft0->is_leader()) {
         // if configuration contains the broker already just update its config
         // with data from join request
 
-        if (_raft0->config().contains_broker(broker.id())) {
+        if (_raft0->config().contains_broker(req.node.id())) {
             vlog(
               clusterlog.info,
               "Broker {} is already member of a cluster, updating "
               "configuration",
-              broker.id());
-            auto req = configuration_update_request(
-              std::move(broker), _self.id());
+              req.node.id());
+            auto node_id = req.node.id();
+            auto update_req = configuration_update_request(
+              req.node, _self.id());
             co_return co_await handle_configuration_update_request(
-              std::move(req))
-              .then([](result<configuration_update_reply> r) {
+              std::move(update_req))
+              .then([node_id](result<configuration_update_reply> r) {
                   if (r) {
-                      return ret_t(join_reply{.success = r.value().success});
+                      auto success = r.value().success;
+                      return ret_t(join_node_reply{
+                        .success = success,
+                        .id = success ? node_id : model::node_id{-1}});
                   }
                   return ret_t(r.error());
               });
         }
 
-        if (_raft0->config().contains_address(broker.rpc_address())) {
+        if (_raft0->config().contains_address(req.node.rpc_address())) {
             vlog(
               clusterlog.info,
               "Broker {} address ({}) conflicts with the address of another "
               "node",
-              broker.id(),
-              broker.rpc_address());
-            co_return ret_t(join_reply{.success = false});
+              req.node.id(),
+              req.node.rpc_address());
+            co_return ret_t(join_node_reply{.success = false});
         }
-        if (broker.id() != _self.id()) {
+        if (req.node.id() != _self.id()) {
             co_await update_broker_client(
               _self.id(),
               _connection_cache,
-              broker.id(),
-              broker.rpc_address(),
+              req.node.id(),
+              req.node.rpc_address(),
               _rpc_tls_config);
         }
         // Just update raft0 configuration
         // we do not use revisions in raft0 configuration, it is always revision
         // 0 which is perfectly fine. this will work like revision less raft
         // protocol.
         co_return co_await _raft0
-          ->add_group_members({broker}, model::revision_id(0))
-          .then([broker](std::error_code ec) {
+          ->add_group_members({req.node}, model::revision_id(0))
+          .then([broker = req.node](std::error_code ec) {
```

Comment:
I think you have use-after-free issue here. In the code below the `SCHEDULING_POINT` causes the coroutine to be suspended and the caller will receive a `ss::future<T>`:

```c++
members_manager::handle_join_request(join_node_request const& req)
        ...
        if (req.node.id() != _self.id()) {
            co_await update_broker_client(    <---- SCHEDULING POINT
              _self.id(),
             ...
        }

        co_return co_await _raft0
          ->add_group_members({req.node}, ... <---- REFERENCE `req`
          .then([broker = req.node](std::error_code ec) {
```

the caller is (nothing problematic here):

```c++
members_manager::dispatch_join_to_seed_server(join_node_request const& req)
    f = handle_join_request(req);
   ...
    return f.then_wrapped([it, this, req](ss::future<ret_t> fut) {
```

but the top of the call stack is is passing a temporary which is correctly kept alive for the duration of the function invocation, but the function returns at the bottom of the call stack at the `SCHEDULING_POINT` so the `REFERENCE` that occurs later will point at the previous temporary value that has been freed:

```c++
void members_manager::join_raft0() {
    ssx::spawn_with_gate(_gate, [this] {
        vlog(clusterlog.debug, "Trying to join the cluster");
        return ss::repeat([this] {
                   return dispatch_join_to_seed_server(
                            std::cbegin(_seed_servers),
                            std::move(join_node_request{.     <--- TEMPORARY
                              feature_table::get_latest_logical_version(),
                              _self}))
```

Suggested fix: coroutines should take values not reference. Although in this case I dunno if this points at some other code paths that need updated too.

simplified reproducer:

```c++
static ss::future<ss::stop_iteration> second(std::string const& s) {
    BOOST_REQUIRE(!s.empty());
    co_await ss::sleep(std::chrono::seconds(1));
    BOOST_REQUIRE(!s.empty());
    co_return ss::stop_iteration::no;
}

static ss::future<> first() {
    return ss::repeat(
      [] { return second(std::string("i am a temporary string")); });
}

SEASTAR_THREAD_TEST_CASE(temp_use_after_free) { first().get(); }
```

Reply:
You're very right.  I've updated handle_join_request to take the request by value.